### DISCUSSION TITLE
frozen_components clarification

### DISCUSSION THREAD

**Discussion by kudryk** [explosion/spaCy#11338] [Help: Other Questions] [answered]
I'd like to clarify the intent of `frozen_components` in the `[training]` section. From the documentation and reading of the code, my understanding is that this prevents any of the named components in that list from being retrained when other components are being trained. Some components extend `TrainablePipe` and those that cannot be trained extend just `Pipe`. Therefore, I presumed that candidates to be included in `frozen_components` should only be those that subclass `TrainablePipe`. I have configured my project to train the `textcat` components under this assumption.

Yet, when I go to train, I see the following line in the output:

`[2022-08-18 20:37:00,795] [INFO] Resuming training for: ['attribute_ruler', 'lemmatizer']`

According to the documentation and source code, these 2 components are not trainable. Why is it resuming training for these 2 components?  When further digging into this, I see right after this message it detects if these components support rehearsing (they don't) but returns an optimizer.

What exactly is going on here? Is it best to let these components be retrained, or should I included them in with `frozen_components` (i.e. my understanding of this field is not correct)?

**Comment by polm** [comment, marked as answer]
Basically, nothing weird is going on here and you can ignore this message.

The components listed in the output you see are just everything that isn't frozen, whether trainable or not - the training loop isn't actually checking the components that closely at this point. It sounds like you looked at [the implementation](https://github.com/explosion/spaCy/blob/cab263791ff25a713bd2a0e72759fa48aff36b9f/spacy/language.py#L1333), but as you mentioned it only does something with the components if they use the optional rehearsal feature (plus some general housekeeping for the pipeline related to the vectors and optimizer). 

Given that, I agree the output here is a little confusing - we'll see if we can improve it. We do want to list what components are running as usual somewhere at the start of training though, and currently this is where that's happening.

**Reply by kudryk** [reply]
Thank you.
